File: test/vectors/verify-checkpoint.mjs
#!/usr/bin/env node
/**
 * Checkpoint conformance vector verifier (JavaScript).
 * Verifies that checkpoint canonicalization, chain hashing, and truncation
 * detection produce byte-identical results across implementations.
 */

import { createHash } from "node:crypto";
import { readFile } from "node:fs/promises";
import { fileURLToPath } from "node:url";
import { dirname, join } from "node:path";

const __dirname = dirname(fileURLToPath(import.meta.url));
const vectors = JSON.parse(await readFile(join(__dirname, "checkpoint.json"), "utf-8"));

let passed = 0;
let failed = 0;

function sha256(input) {
  return createHash("sha256").update(input).digest("hex");
}

function canonicalizeCheckpoint(record) {
  const ordered = [
    ["id", record.id],
    ["type", "checkpoint"],
    ["timestamp", record.timestamp],
    ["sequence", record.sequence],
    ["recordCount", record.recordCount],
    ["previousHash", record.previousHash],
  ];
  if (record.parties != null) {
    ordered.push(["parties", record.parties]);
  }
  return JSON.stringify(ordered);
}

function hashRecord(record) {
  return createHash("sha256").update(JSON.stringify(record)).digest("hex");
}

function assert(condition, name, detail) {
  if (condition) {
    passed++;
    console.log(`  PASS: ${name}`);
  } else {
    failed++;
    console.log(`  FAIL: ${name}`);
    if (detail) console.log(`        ${detail}`);
  }
}

// --- Checkpoint Canonicalization ---
console.log("\n=== Checkpoint Canonicalization ===");
for (const vec of vectors.checkpoint_canonicalization) {
  const canonical = canonicalizeCheckpoint(vec.record);
  assert(
    canonical === vec.canonical,
    `${vec.name} canonical form`,
    `expected: ${vec.canonical}\n        got:      ${canonical}`
  );
  const hash = sha256(canonical);
  assert(
    hash === vec.sha256_canonical,
    `${vec.name} SHA-256`,
    `expected: ${vec.sha256_canonical}\n        got:      ${hash}`
  );
}

// --- Checkpoint Chain ---
console.log("\n=== Checkpoint Chain ===");
const chainRecords = vectors.checkpoint_chain.records;
for (let i = 0; i < chainRecords.length; i++) {
  const entry = chainRecords[i];
  const computedHash = hashRecord(entry.record);
  assert(
    computedHash === entry.record_hash,
    `chain record ${i} hash`,
    `expected: ${entry.record_hash}\n        got:      ${computedHash}`
  );

  if (i > 0) {
    assert(
      entry.record.previousHash === chainRecords[i - 1].record_hash,
      `chain record ${i} previousHash links to record ${i - 1}`,
      `expected: ${chainRecords[i - 1].record_hash}\n        got:      ${entry.record.previousHash}`
    );
  }
}

// --- Truncation Detection ---
console.log("\n=== Truncation Detection ===");
const truncVec = vectors.truncation_detection;
const extCkpt = truncVec.external_checkpoint;

// Full chain should contain the checkpoint
const fullChain = chainRecords.map(e => e.record);
let foundInFull = false;
for (const rec of fullChain) {
  if (rec.type === "checkpoint" &&
      rec.previousHash === extCkpt.previousHash &&
      rec.sequence === extCkpt.sequence &&
      rec.recordCount === extCkpt.recordCount) {
    foundInFull = true;
  }
}
assert(foundInFull, "full chain contains externalized checkpoint");

// Truncated chain should NOT contain the checkpoint
const truncatedChain = truncVec.truncated_chain.records_delivered;
let foundInTruncated = false;
let hasDescendant = false;
for (const rec of truncatedChain) {
  if (rec.type === "checkpoint") {
    if (rec.previousHash === extCkpt.previousHash &&
        rec.sequence === extCkpt.sequence &&
        rec.recordCount === extCkpt.recordCount) {
      foundInTruncated = true;
    }
    if (rec.sequence > extCkpt.sequence) {
      hasDescendant = true;
    }
  }
}
assert(!foundInTruncated && !hasDescendant, "truncated chain missing checkpoint (truncation detected)");

// --- canonicalizeValue ---
console.log("\n=== canonicalizeValue ===");

function canonicalizeValue(value) {
  if (value === null || value === undefined) return null;
  switch (typeof value) {
    case "string":
      for (let i = 0; i < value.length; i++) {
        const code = value.charCodeAt(i);
        if (code >= 0xD800 && code <= 0xDBFF) {
          const next = i + 1 < value.length ? value.charCodeAt(i + 1) : 0;
          if (next < 0xDC00 || next > 0xDFFF)
            throw new Error(`unpaired surrogate at index ${i}`);
          i++;
        } else if (code >= 0xDC00 && code <= 0xDFFF) {
          throw new Error(`unpaired surrogate at index ${i}`);
        }
      }
      return value;
    case "boolean":
      return value;
    case "number":
      if (!Number.isSafeInteger(value))
        throw new Error(`unsafe number ${value}`);
      return value;
    case "object":
      if (Array.isArray(value)) return ["L", value.map(canonicalizeValue)];
      const keys = Object.keys(value).sort().filter(k => value[k] !== undefined);
      return ["M", keys.map(k => [k, canonicalizeValue(value[k])])];
    default:
      throw new Error(`unsupported type ${typeof value}`);
  }
}

function computeExtensionsDigest(extensions) {
  const canonicalized = canonicalizeValue(extensions);
  const canonical = JSON.stringify(canonicalized);
  return { canonical, digest: sha256(canonical) };
}

const cvVectors = vectors.canonicalize_value.vectors;
for (const vec of cvVectors) {
  if (vec.expected_error) {
    if (vec.construct) {
      // Programmatic construction (e.g. lone surrogate can't be in JSON)
      let threw = false;
      try { canonicalizeValue(String.fromCharCode(0xD800)); } catch { threw = true; }
      assert(threw, `${vec.name} throws on invalid input`);
    } else {
      let threw = false;
      try { canonicalizeValue(vec.input); } catch { threw = true; }
      assert(threw, `${vec.name} throws on invalid input`);
    }
    continue;
  }
  if (vec.input_a && vec.input_b && vec.canonical_form) {
    // Same content different order -> same digest
    const ra = computeExtensionsDigest(vec.input_a);
    const rb = computeExtensionsDigest(vec.input_b);
    assert(ra.canonical === vec.canonical_form, `${vec.name} canonical form`,
      `expected: ${vec.canonical_form}\n        got:      ${ra.canonical}`);
    assert(ra.digest === vec.digest, `${vec.name} digest`, `expected: ${vec.digest}\n        got:      ${ra.digest}`);
    assert(ra.digest === rb.digest, `${vec.name} both inputs produce same digest`);
  } else if (vec.input_a && vec.input_b && vec.canonical_a) {
    // Array order matters
    const ra = computeExtensionsDigest(vec.input_a);
    const rb = computeExtensionsDigest(vec.input_b);
    assert(ra.canonical === vec.canonical_a, `${vec.name} canonical_a`);
    assert(ra.digest === vec.digest_a, `${vec.name} digest_a`);
    assert(rb.canonical === vec.canonical_b, `${vec.name} canonical_b`);
    assert(rb.digest === vec.digest_b, `${vec.name} digest_b`);
    assert(ra.digest !== rb.digest, `${vec.name} digests differ`);
  } else if (vec.input) {
    // Unicode keys
    const r = computeExtensionsDigest(vec.input);
    assert(r.canonical === vec.canonical_form, `${vec.name} canonical form`,
      `expected: ${vec.canonical_form}\n        got:      ${r.canonical}`);
    assert(r.digest === vec.digest, `${vec.name} digest`,
      `expected: ${vec.digest}\n        got:      ${r.digest}`);
  }
}

// --- Extensions Digest ---
console.log("\n=== Extensions Digest ===");
const extVectors = vectors.extensions_digest.vectors;

for (const vec of extVectors) {
  const { canonical, digest } = computeExtensionsDigest(vec.extensions);
  assert(
    canonical === vec.canonical_form,
    `${vec.name} canonical form`,
    `expected: ${vec.canonical_form}\n        got:      ${canonical}`
  );
  assert(
    digest === vec.digest,
    `${vec.name} digest`,
    `expected: ${vec.digest}\n        got:      ${digest}`
  );
}

// Record canonicalization with extensionsDigest
function canonicalizeRecord(record) {
  const ordered = [
    ["id", record.id],
    ["timestamp", record.timestamp],
    ["method", record.method],
    ["toolName", record.toolName ?? null],
    ["namespace", record.namespace ?? null],
    ["upstream", record.upstream ?? null],
    ["principal", record.principal ?? null],
    ["durationMs", record.durationMs],
    ["success", record.success],
    ["errorCode", record.errorCode ?? null],
    ["previousHash", record.previousHash ?? null],
  ];
  let insertAt = 11;
  if (record.decisionContextDigest != null) {
    ordered.splice(10, 0, ["decisionContextDigest", record.decisionContextDigest]);
    insertAt = 12;
  }
  if (record.extensionsDigest != null) {
    ordered.splice(insertAt, 0, ["extensionsDigest", record.extensionsDigest]);
    insertAt++;
  }
  if (record.parties != null) {
    ordered.splice(insertAt, 0, ["parties", record.parties]);
  }
  return JSON.stringify(ordered);
}

const withExt = vectors.extensions_digest.record_canonicalization.with_extensions_digest;
const withExtCanonical = canonicalizeRecord(withExt.record);
assert(
  withExtCanonical === withExt.canonical,
  "record with extensionsDigest canonical form",
  `expected: ${withExt.canonical}\n        got:      ${withExtCanonical}`
);
assert(
  sha256(withExtCanonical) === withExt.sha256_canonical,
  "record with extensionsDigest SHA-256",
  `expected: ${withExt.sha256_canonical}\n        got:      ${sha256(withExtCanonical)}`
);

const withoutExt = vectors.extensions_digest.record_canonicalization.without_extensions_digest;
const withoutExtCanonical = canonicalizeRecord(withoutExt.record);
assert(
  withoutExtCanonical === withoutExt.canonical,
  "record without extensionsDigest canonical form (backward compat)",
  `expected: ${withoutExt.canonical}\n        got:      ${withoutExtCanonical}`
);
assert(
  sha256(withoutExtCanonical) === withoutExt.sha256_canonical,
  "record without extensionsDigest SHA-256",
  `expected: ${withoutExt.sha256_canonical}\n        got:      ${sha256(withoutExtCanonical)}`
);

// --- Rotation Boundary ---
console.log("\n=== Rotation Boundary ===");
const rotation = vectors.rotation_boundary;

// Verify file 1 hashes
for (let i = 0; i < rotation.file_1_records.length; i++) {
  const entry = rotation.file_1_records[i];
  const h = hashRecord(entry.record);
  assert(h === entry.record_hash, `rotation file1 record ${i} hash`,
    `expected: ${entry.record_hash}\n        got:      ${h}`);
}

// Verify file 2 chains to file 1's last record
const file1LastHash = rotation.file_1_records[rotation.file_1_records.length - 1].record_hash;
const file2First = rotation.file_2_records[0];
assert(
  file2First.record.previousHash === file1LastHash,
  "rotation: file2 first record chains to file1 last hash",
  `expected: ${file1LastHash}\n        got:      ${file2First.record.previousHash}`
);
const file2Hash = hashRecord(file2First.record);
assert(file2Hash === file2First.record_hash, "rotation file2 record 0 hash",
  `expected: ${file2First.record_hash}\n        got:      ${file2Hash}`);

// --- Sequence Regression ---
console.log("\n=== Sequence Regression ===");
const seqReg = vectors.sequence_regression;
const checkpoints = seqReg.chain.filter(e => e.record.type === "checkpoint");
let regressionDetected = false;
for (let i = 1; i < checkpoints.length; i++) {
  if (checkpoints[i].record.sequence <= checkpoints[i - 1].record.sequence) {
    regressionDetected = true;
  }
}
assert(regressionDetected, "sequence regression detected in chain");
assert(
  seqReg.detection_result.failureCode === "sequence_regression",
  "failure code is sequence_regression"
);

// --- Chain Break ---
console.log("\n=== Chain Break ===");
const chainBreak = vectors.chain_break;
for (let i = 0; i < chainBreak.records.length; i++) {
  const entry = chainBreak.records[i];
  const h = hashRecord(entry.record);
  assert(h === entry.record_hash, `chain_break record ${i} hash`,
    `expected: ${entry.record_hash}\n        got:      ${h}`);
}
// Verify chaining: second record's previousHash = first record's hash
assert(
  chainBreak.records[1].record.previousHash === chainBreak.records[0].record_hash,
  "record after chain_break chains from break record hash"
);

// --- Summary ---
console.log(`\n=== Results: ${passed} passed, ${failed} failed (${passed + failed} total) ===`);
process.exit(failed > 0 ? 1 : 0);

File: test/vectors/verify-checkpoint.py
#!/usr/bin/env python3
"""
Checkpoint conformance vector verifier (Python).
Verifies that checkpoint canonicalization, chain hashing, and truncation
detection produce byte-identical results across implementations.
"""

import hashlib
import json
import sys
from pathlib import Path

vectors_path = Path(__file__).parent / "checkpoint.json"
vectors = json.loads(vectors_path.read_text())

passed = 0
failed = 0


def sha256(data: str) -> str:
    return hashlib.sha256(data.encode("utf-8")).hexdigest()


def assert_well_formed_string(value: str, context: str) -> None:
    for i, ch in enumerate(value):
        code = ord(ch)
        if 0xD800 <= code <= 0xDFFF:
            raise ValueError(f"{context}: unpaired surrogate at index {i}")


def canonicalize_checkpoint(record: dict) -> str:
    assert_well_formed_string(record["id"], "canonicalizeCheckpoint.id")
    assert_well_formed_string(record["timestamp"], "canonicalizeCheckpoint.timestamp")
    assert_well_formed_string(record["previousHash"], "canonicalizeCheckpoint.previousHash")
    ordered = [
        ["id", record["id"]],
        ["type", "checkpoint"],
        ["timestamp", record["timestamp"]],
        ["sequence", record["sequence"]],
        ["recordCount", record["recordCount"]],
        ["previousHash", record["previousHash"]],
    ]
    if record.get("parties") is not None:
        ordered.append(["parties", record["parties"]])
    return json.dumps(ordered, separators=(",", ":"), ensure_ascii=False)


def hash_record(record: dict) -> str:
    serialized = json.dumps(record, separators=(",", ":"), ensure_ascii=False)
    return hashlib.sha256(serialized.encode("utf-8")).hexdigest()


def check(condition: bool, name: str, detail: str = ""):
    global passed, failed
    if condition:
        passed += 1
        print(f"  PASS: {name}")
    else:
        failed += 1
        print(f"  FAIL: {name}")
        if detail:
            print(f"        {detail}")


# --- Checkpoint Canonicalization ---
print("\n=== Checkpoint Canonicalization ===")
for vec in vectors["checkpoint_canonicalization"]:
    canonical = canonicalize_checkpoint(vec["record"])
    check(
        canonical == vec["canonical"],
        f"{vec['name']} canonical form",
        f"expected: {vec['canonical']}\n        got:      {canonical}",
    )
    h = sha256(canonical)
    check(
        h == vec["sha256_canonical"],
        f"{vec['name']} SHA-256",
        f"expected: {vec['sha256_canonical']}\n        got:      {h}",
    )

# --- Checkpoint Chain ---
print("\n=== Checkpoint Chain ===")
chain_records = vectors["checkpoint_chain"]["records"]
for i, entry in enumerate(chain_records):
    computed = hash_record(entry["record"])
    check(
        computed == entry["record_hash"],
        f"chain record {i} hash",
        f"expected: {entry['record_hash']}\n        got:      {computed}",
    )
    if i > 0:
        check(
            entry["record"]["previousHash"] == chain_records[i - 1]["record_hash"],
            f"chain record {i} previousHash links to record {i - 1}",
            f"expected: {chain_records[i - 1]['record_hash']}\n        got:      {entry['record']['previousHash']}",
        )

# --- Truncation Detection ---
print("\n=== Truncation Detection ===")
trunc_vec = vectors["truncation_detection"]
ext_ckpt = trunc_vec["external_checkpoint"]

# Full chain should contain the checkpoint
full_chain = [e["record"] for e in chain_records]
found_in_full = any(
    r.get("type") == "checkpoint"
    and r.get("previousHash") == ext_ckpt["previousHash"]
    and r.get("sequence") == ext_ckpt["sequence"]
    and r.get("recordCount") == ext_ckpt["recordCount"]
    for r in full_chain
)
check(found_in_full, "full chain contains externalized checkpoint")

# Truncated chain should NOT contain the checkpoint
truncated_chain = trunc_vec["truncated_chain"]["records_delivered"]
found_in_truncated = any(
    r.get("type") == "checkpoint"
    and r.get("previousHash") == ext_ckpt["previousHash"]
    and r.get("sequence") == ext_ckpt["sequence"]
    and r.get("recordCount") == ext_ckpt["recordCount"]
    for r in truncated_chain
)
has_descendant = any(
    r.get("type") == "checkpoint" and r.get("sequence", 0) > ext_ckpt["sequence"]
    for r in truncated_chain
)
check(
    not found_in_truncated and not has_descendant,
    "truncated chain missing checkpoint (truncation detected)",
)

# --- canonicalizeValue ---
print("\n=== canonicalizeValue ===")


def canonicalize_value(value):
    if value is None:
        return None
    if isinstance(value, bool):
        return value
    if isinstance(value, int):
        if abs(value) > 2**53 - 1:
            raise ValueError(f"unsafe number {value}")
        return value
    if isinstance(value, float):
        raise ValueError(f"unsafe number {value}")
    if isinstance(value, str):
        for i, ch in enumerate(value):
            code = ord(ch)
            if 0xD800 <= code <= 0xDFFF:
                raise ValueError(f"unpaired surrogate at index {i}")
        return value
    if isinstance(value, list):
        return ["L", [canonicalize_value(v) for v in value]]
    if isinstance(value, dict):
        # Sort by UTF-16 code-unit order to match JavaScript's String.prototype.sort().
        # Python sorts by code points; these diverge for astral-plane characters.
        keys = sorted(value.keys(), key=lambda k: k.encode("utf-16-be"))
        return ["M", [[k, canonicalize_value(value[k])] for k in keys]]
    raise ValueError(f"unsupported type {type(value)}")


def compute_extensions_digest(extensions: dict) -> tuple:
    canonicalized = canonicalize_value(extensions)
    canonical = json.dumps(canonicalized, separators=(",", ":"), ensure_ascii=False)
    return canonical, sha256(canonical)


cv_vectors = vectors["canonicalize_value"]["vectors"]
for vec in cv_vectors:
    if "expected_error" in vec:
        threw = False
        if "construct" in vec:
            # Programmatic: lone surrogate can't reliably live in JSON
            try:
                canonicalize_value(chr(0xD800))
            except (ValueError, UnicodeEncodeError):
                threw = True
        else:
            try:
                canonicalize_value(vec["input"])
            except ValueError:
                threw = True
        check(threw, f"{vec['name']} throws on invalid input")
        continue
    if "input_a" in vec and "input_b" in vec and "canonical_form" in vec:
        ra_c, ra_d = compute_extensions_digest(vec["input_a"])
        rb_c, rb_d = compute_extensions_digest(vec["input_b"])
        check(ra_c == vec["canonical_form"], f"{vec['name']} canonical form",
              f"expected: {vec['canonical_form']}\n        got:      {ra_c}")
        check(ra_d == vec["digest"], f"{vec['name']} digest",
              f"expected: {vec['digest']}\n        got:      {ra_d}")
        check(ra_d == rb_d, f"{vec['name']} both inputs produce same digest")
    elif "input_a" in vec and "canonical_a" in vec:
        ra_c, ra_d = compute_extensions_digest(vec["input_a"])
        rb_c, rb_d = compute_extensions_digest(vec["input_b"])
        check(ra_c == vec["canonical_a"], f"{vec['name']} canonical_a")
        check(ra_d == vec["digest_a"], f"{vec['name']} digest_a")
        check(rb_c == vec["canonical_b"], f"{vec['name']} canonical_b")
        check(rb_d == vec["digest_b"], f"{vec['name']} digest_b")
        check(ra_d != rb_d, f"{vec['name']} digests differ")
    elif "input" in vec:
        r_c, r_d = compute_extensions_digest(vec["input"])
        check(r_c == vec["canonical_form"], f"{vec['name']} canonical form",
              f"expected: {vec['canonical_form']}\n        got:      {r_c}")
        check(r_d == vec["digest"], f"{vec['name']} digest",
              f"expected: {vec['digest']}\n        got:      {r_d}")

# --- Extensions Digest ---
print("\n=== Extensions Digest ===")
ext_vectors = vectors["extensions_digest"]["vectors"]

for vec in ext_vectors:
    canonical, digest = compute_extensions_digest(vec["extensions"])
    check(
        canonical == vec["canonical_form"],
        f"{vec['name']} canonical form",
        f"expected: {vec['canonical_form']}\n        got:      {canonical}",
    )
    check(
        digest == vec["digest"],
        f"{vec['name']} digest",
        f"expected: {vec['digest']}\n        got:      {digest}",
    )


def canonicalize_record(record: dict) -> str:
    assert_well_formed_string(record["id"], "canonicalizeRecord.id")
    assert_well_formed_string(record["timestamp"], "canonicalizeRecord.timestamp")
    assert_well_formed_string(record["method"], "canonicalizeRecord.method")
    if record.get("toolName") is not None:
        assert_well_formed_string(record["toolName"], "canonicalizeRecord.toolName")
    if record.get("namespace") is not None:
        assert_well_formed_string(record["namespace"], "canonicalizeRecord.namespace")
    if record.get("upstream") is not None:
        assert_well_formed_string(record["upstream"], "canonicalizeRecord.upstream")
    if record.get("principal") is not None:
        assert_well_formed_string(record["principal"], "canonicalizeRecord.principal")
    if record.get("previousHash") is not None:
        assert_well_formed_string(record["previousHash"], "canonicalizeRecord.previousHash")
    if record.get("decisionContextDigest") is not None:
        assert_well_formed_string(record["decisionContextDigest"], "canonicalizeRecord.decisionContextDigest")
    if record.get("extensionsDigest") is not None:
        assert_well_formed_string(record["extensionsDigest"], "canonicalizeRecord.extensionsDigest")
    ordered = [
        ["id", record["id"]],
        ["timestamp", record["timestamp"]],
        ["method", record["method"]],
        ["toolName", record.get("toolName")],
        ["namespace", record.get("namespace")],
        ["upstream", record.get("upstream")],
        ["principal", record.get("principal")],
        ["durationMs", record["durationMs"]],
        ["success", record["success"]],
        ["errorCode", record.get("errorCode")],
        ["previousHash", record.get("previousHash")],
    ]
    insert_at = 11
    if record.get("decisionContextDigest") is not None:
        ordered.insert(10, ["decisionContextDigest", record["decisionContextDigest"]])
        insert_at = 12
    if record.get("extensionsDigest") is not None:
        ordered.insert(insert_at, ["extensionsDigest", record["extensionsDigest"]])
        insert_at += 1
    if record.get("aiInvocation") is not None:
        ordered.insert(insert_at, ["aiInvocation", canonicalize_value(record["aiInvocation"])])
        insert_at += 1
    if record.get("parties") is not None:
        ordered.insert(insert_at, ["parties", record["parties"]])
    return json.dumps(ordered, separators=(",", ":"), ensure_ascii=False)


rec_canon = vectors["extensions_digest"]["record_canonicalization"]

with_ext = rec_canon["with_extensions_digest"]
with_ext_canonical = canonicalize_record(with_ext["record"])
check(
    with_ext_canonical == with_ext["canonical"],
    "record with extensionsDigest canonical form",
    f"expected: {with_ext['canonical']}\n        got:      {with_ext_canonical}",
)
check(
    sha256(with_ext_canonical) == with_ext["sha256_canonical"],
    "record with extensionsDigest SHA-256",
    f"expected: {with_ext['sha256_canonical']}\n        got:      {sha256(with_ext_canonical)}",
)

without_ext = rec_canon["without_extensions_digest"]
without_ext_canonical = canonicalize_record(without_ext["record"])
check(
    without_ext_canonical == without_ext["canonical"],
    "record without extensionsDigest canonical form (backward compat)",
    f"expected: {without_ext['canonical']}\n        got:      {without_ext_canonical}",
)
check(
    sha256(without_ext_canonical) == without_ext["sha256_canonical"],
    "record without extensionsDigest SHA-256",
    f"expected: {without_ext['sha256_canonical']}\n        got:      {sha256(without_ext_canonical)}",
)

# --- Rotation Boundary ---
print("\n=== Rotation Boundary ===")
rotation = vectors["rotation_boundary"]

for i, entry in enumerate(rotation["file_1_records"]):
    h = hash_record(entry["record"])
    check(
        h == entry["record_hash"],
        f"rotation file1 record {i} hash",
        f"expected: {entry['record_hash']}\n        got:      {h}",
    )

file1_last_hash = rotation["file_1_records"][-1]["record_hash"]
file2_first = rotation["file_2_records"][0]
check(
    file2_first["record"]["previousHash"] == file1_last_hash,
    "rotation: file2 first record chains to file1 last hash",
    f"expected: {file1_last_hash}\n        got:      {file2_first['record']['previousHash']}",
)
file2_hash = hash_record(file2_first["record"])
check(
    file2_hash == file2_first["record_hash"],
    "rotation file2 record 0 hash",
    f"expected: {file2_first['record_hash']}\n        got:      {file2_hash}",
)

# --- Sequence Regression ---
print("\n=== Sequence Regression ===")
seq_reg = vectors["sequence_regression"]
checkpoints = [e for e in seq_reg["chain"] if e["record"].get("type") == "checkpoint"]
regression_detected = False
for i in range(1, len(checkpoints)):
    if checkpoints[i]["record"]["sequence"] <= checkpoints[i - 1]["record"]["sequence"]:
        regression_detected = True
check(regression_detected, "sequence regression detected in chain")
check(
    seq_reg["detection_result"]["failureCode"] == "sequence_regression",
    "failure code is sequence_regression",
)

# --- Chain Break ---
print("\n=== Chain Break ===")
chain_break = vectors["chain_break"]
for i, entry in enumerate(chain_break["records"]):
    h = hash_record(entry["record"])
    check(
        h == entry["record_hash"],
        f"chain_break record {i} hash",
        f"expected: {entry['record_hash']}\n        got:      {h}",
    )
check(
    chain_break["records"][1]["record"]["previousHash"] == chain_break["records"][0]["record_hash"],
    "record after chain_break chains from break record hash",
)

# --- Summary ---
print(f"\n=== Results: {passed} passed, {failed} failed ({passed + failed} total) ===")
sys.exit(1 if failed > 0 else 0)
